src/osd/ReplicatedPG.cc

```diff
@@ -3701,6 +3701,40 @@ int ReplicatedPG::do_xattr_cmp_str(int op, string& v1s, bufferlist& xattr)
   }
 }
 
+int ReplicatedPG::do_writesame(OpContext *ctx, OSDOp& osd_op)
+{
+  ceph_osd_op& op = osd_op.op;
+  vector<OSDOp> write_ops(1);
+  OSDOp& write_op = write_ops[0];
+  uint64_t write_length = op.writesame.length;
+  int result = 0;
+
+  if (!write_length)
+    return 0;
+
+  if (!op.writesame.data_length || write_length % op.writesame.data_length)
+    return -EINVAL;
+
+  if (op.writesame.data_length != osd_op.indata.length()) {
+    derr << "invalid length ws data length " << op.writesame.data_length << " actual len " << osd_op.indata.length() << dendl;
+    return -EINVAL;
+  }
+
+  while (write_length) {
+    write_op.indata.append(osd_op.indata.c_str(), op.writesame.data_length);
+    write_length -= op.writesame.data_length;
+  }
+
+  write_op.op.op = CEPH_OSD_OP_WRITE;
```

Comment:
@jdurgin: Exactly.

@yuyuyu101: The (current) use-case for CEPH_OSD_OP_WRITESAME is to offload handling of SCSI WRITE SAME I/O to the OSDs.
SCSI WRITE SAME can be used by a SCSI initiator in a couple of ways:
1) write an arbitrary buffer multiple times to disk
2) trim/discard a range on a thin-provisioned disk (UNMAP=1)

For type (2) WRITE SAME handling, the kernel client uses CEPH_OSD_OP_DELETE, CEPH_OSD_OP_TRUNCATE or CEPH_OSD_OP_ZERO to implement the trim/discard behaviour.

For type (1) WRITE SAME handling, CEPH_OSD_OP_WRITESAME will initially be used. CEPH_OSD_OP_ZERO could be used later to optimise the case where the buffer is all zeros.